src/miniscript/iter.rs

```diff
@@ -0,0 +1,552 @@
+impl<Pk: MiniscriptKey, Ctx: ScriptContext> Miniscript<Pk, Ctx> {
+    /// Creates a new [Iter] iterator that will iterate over all [Miniscript] items within
+    /// AST by traversing its branches. For the specific algorithm please see
+    /// [Iter::next] function.
+    pub fn iter(&self) -> Iter<Pk, Ctx> {
+        Iter::new(self)
+    }
+
+    /// Creates a new [PkIter] iterator that will iterate over all plain public keys (and not
+    /// key hash values) present in [Miniscript] items within AST by traversing all its branches.
+    /// For the specific algorithm please see [PkIter::next] function.
+    pub fn iter_pk(&self) -> PkIter<Pk, Ctx> {
+        PkIter::new(self)
+    }
+
+    /// Creates a new [PkhIter] iterator that will iterate over all public keys hashes (and not
+    /// plain public keys) present in Miniscript items within AST by traversing all its branches.
+    /// For the specific algorithm please see [PkhIter::next] function.
+    pub fn iter_pkh(&self) -> PkhIter<Pk, Ctx> {
+        PkhIter::new(self)
+    }
+
+    /// Creates a new [PkPkhIter] iterator that will iterate over all plain public keys and
+    /// key hash values present in Miniscript items within AST by traversing all its branches.
+    /// For the specific algorithm please see [PkPkhIter::next] function.
+    pub fn iter_pk_pkh(&self) -> PkPkhIter<Pk, Ctx> {
+        PkPkhIter::new(self)
+    }
+
+    /// Enumerates all child nodes of the current AST node (`self`) and returns a `Vec` referencing
+    /// them.
+    pub fn branches(&self) -> Vec<&Miniscript<Pk, Ctx>> {
+        use Terminal::*;
+        match &self.node {
+            &PkK(_) | &PkH(_) | &Multi(_, _) => vec![],
+
+            &Alt(ref node)
+            | &Swap(ref node)
+            | &Check(ref node)
+            | &DupIf(ref node)
+            | &Verify(ref node)
+            | &NonZero(ref node)
+            | &ZeroNotEqual(ref node) => vec![node.deref()],
+
+            &AndV(ref node1, ref node2)
+            | &AndB(ref node1, ref node2)
+            | &OrB(ref node1, ref node2)
+            | &OrD(ref node1, ref node2)
+            | &OrC(ref node1, ref node2)
+            | &OrI(ref node1, ref node2) => vec![node1.deref(), node2.deref()],
+
+            &AndOr(ref node1, ref node2, ref node3) => {
+                vec![node1.deref(), node2.deref(), node3.deref()]
+            }
+
+            &Thresh(_, ref node_vec) => node_vec.iter().map(Arc::deref).collect(),
+
+            _ => vec![],
+        }
+    }
+
+    /// Returns `Vec` with cloned version of all public keys from the current miniscript item,
+    /// if any. Otherwise returns an empty `Vec`.
+    ///
+    /// NB: The function analyzes only single miniscript item and not any of its descendants in AST.
+    /// To obtain a list of all public keys within AST use [`iter_pk()`] function, for example
+    /// `miniscript.iter_pubkeys().collect()`.
+    pub fn get_leaf_pk(&self) -> Vec<Pk> {
+        match self.node.clone() {
```

Comment:
Why do we unconditionally clone here even though for most nodes we just throw away the result?

Reply:
It's also unclear to me what the value is in making this method be public.

Reply:
Function is public since it may be useful. It is different from iterator methods: iterators traverse the tree, while this function returns keys only from the current node.